marklogic/rows.py

```diff
@@ -40,4 +41,113 @@ def graphql(self, graphql_query, return_response=False, *args, **kwargs):
+        Just like that endpoint, this function can be used for four different types of
+        queries: Optic DSL, Serialized Optic, SQL, and SPARQL. The type of query
+        processed by the function is dependent upon the parameter used in the call to
+        the function.
+        For more information about Optic and using the Optic DSL, SQL, and SPARQL,
+        see https://docs.marklogic.com/guide/app-dev/OpticAPI
+        If multiple query parameters are passed into the call, the function uses the
+        query parameter that is first in the list: dsl, plan, sql, sparql.
+
+        :param dsl: an Optic DSL query
+        :param plan: a serialized Optic query
+        :param sql: an SQL query
+        :param sparql: a SPARQL query
+        :param return_response: boolean specifying if the entire original response
+        object should be returned (True) or if only the data should be returned (False)
+        upon a success (2xx) response. Note that if the status code of the response is
+        not 2xx, then the entire response is always returned.
+        """
+        request_info = self.__get_request_info(dsl, plan, sql, sparql)
+        headers = kwargs.pop("headers", {})
+        headers["Content-Type"] = request_info["content-type"]
+        headers["Accept"] = self.__get_accept_header(format)
+        response = self._session.post(
+            "v1/rows",
+            headers=headers,
+            data=request_info["data"],
+            **kwargs
+        )
+        return (
+            self.__get_response_data(response, format)
+            if response.status_code == 200 and not return_response
+            else response
+        )
+
+    def __get_request_info(self, dsl, plan, sql, sparql):
+        """
+        Examine the parameters passed into the query function to determine what value
+        should be passed to the endpoint and what the content-type header should be.
+
+        :param dsl: an Optic DSL query
+        :param plan: a serialized Optic query
+        :param sql: an SQL query
+        :param sparql: a SPARQL query
+        dict object returned contains the two values required to make the POST request.
+        """
+        if dsl is not None:
+            return {
+                "content-type": "application/vnd.marklogic.querydsl+javascript",
+                "data": dsl
+            }
+        if plan is not None:
+            return {
+                "content-type": "application/json",
+                "data": plan
+            }
+        if sql is not None:
+            return {
+                "content-type": "application/sql",
+                "data": sql
+            }
+        if sparql is not None:
+            return {
+                "content-type": "application/sparql-query",
+                "data": sparql
+            }
+        else:
+            raise ValueError(
+                "A valid request parameter must be included. One of: dsl, plan, sql, or sparql")
```

Comment:
Slight tweak for the error message - the user is passing in an arg, not a request parameter. I think you can remove that expression though and just say "Must specify one of: dsl, plan, ..." or perhaps "No query found; must specify one of: dsl, plan, ...".